Make switch a state machine

jacobheun opened the pull request:
Getting this pushed up for visibility while I work on cleanup and add some additional tests.

## Goal
This PR updates libp2p-switch to be a state machine, including its connections #24. 

## Updates
* Switch will now emit `start`, `stop`, and `error` events. Callback legacy support is included.
* Incoming connections will now ensure crypto is done before other muxing. Resolves #239 .
* Outgoing and Incoming connections are now state machines. Rather than being internally, automatically driven, they are driven by dialer.js and handler.js respectively. This allows finer control over transitions and helps provide clarity around when and where errors occur.
* Adds in some error standardization via the `err-codes` module
* Fixed readme links

## TODOs
- [x] There is redundant logic with the new Connection FSMs that needs to be cleaned up.
- [x] Add more, thorough tests for each of the connection FSMs.
- [x] Ensure incoming connections are being properly reused
- [x] General TODO cleanup
- [x] Add `class-is` module and remove instanceof logic, as it's unreliable across package versions.

## Thoughts on Future Additions
* As error codes are added throughout the system, have the FSMs handle fatal and non fatal errors appropriately.
* Add hooks for new connections to allow modules like the connection manager, to make intelligent decisions about those connections (dial backoffs, etc).

## References
* [Libp2p State Machine#98](https://github.com/libp2p/js-libp2p/issues/98)

Required by https://github.com/libp2p/js-libp2p/pull/257

jacobheun commented:
@diasdavid @pgte @vasco-santos I've still got some cleanup to do per my notes in the PR overview, but I'd love to get some feedback on the overall implementation of the FSM. There may be some considerations I've missed that we could roll into this update.

pgte commented on src/connection/index.js:
```diff
@@ -0,0 +1,456 @@
+        if (circuitTried) {
+          this.emit('error', new Error(`No available transports to dial peer ${this.theirB58Id}!`))
+          return this._state('disconnect')
+        }
```
These last two conditions can dictate the end of the dialing attempt. As a client of the switch API, I'd like to now the reason why a dial failed. "No available transport" or "Circuit not enabled" doesn't tell me much of why the other transports failed.

As a remedy, I suggest that we keep track of the failing reasons and make them somehow available to the switch user.

pgte replied:
As an alternative, emit a warning-like event like "connection attempt failed", containing an error object as payload.

jacobheun replied:
I like the warning option as that would provide the user with more information about the various attempts, even if connections succeed on a later transport.

With the current structure the switch would need to emit the events. Perhaps the payload should also include the peerInfo so users understand where it came from. Something like:
```js
_switch.emit('error:connection_attempt_failed', {
  peerInfo: theirPeerInfo,
  error: ...
})
```

pgte replied:
I see. This dialing is triggered by a `switch.dial()`, which is callback-based..
Having the switch emit these events independently of the dial() call means that, for the API user,  it will be hard to correlate a call with these warnings.
Usually I solve this by either returning an even stream or an event emitter. In this case, I think that something like a dialFSM would be something useful to consider...

jacobheun replied:
That's a good point. Currently `Libp2p` doesn't return anything from its dial methods so returning an event emitter there from a `Switch` dialFSM would enable us to keep backwards compatibility more easily while we transition the system.

For js-ipfs we'd also need to make sure to return that emitter from [`swarm.connect()`](https://github.com/ipfs/js-ipfs/blob/c73bd2f68a57651e2b7a67b522f79556040963ae/src/core/components/swarm.js#L71).

jacobheun replied:
@pgte I've added a `dialsFSM` method to the switch, that can be used instead of `dial`. Instead of a connection, it will return the fsm. I've added some tests for it to show the events and functionality it has https://github.com/libp2p/js-libp2p-switch/pull/278/files#diff-080e21482eea9ffdd7ee511eb71d3522.

**The main things to note**:
The raw connection is still returned in the callback passed to dialFSM
1. Users can listen on `error` events to get fatal messages
2. Users can listen on `error:connection_attempt_failed` events to get a list of errors for failed transport dials, these are not fatal to the connection.
3. Users can listen on `close` events to know when the specific connection is closed.
4. Users can call `fsm.close()` to close that specific connection.

Are there other events or methods that would be helpful to have? We can always wait to add other things if there aren't higher priority nice to haves.

pgte replied:
@jacobheun That's great! Other than successful connection attempt (which you can infer from having the "connected" callback called), I don't see any other so far.

pgte commented on src/connection/index.js:
```diff
@@ -0,0 +1,456 @@
+          })
+
+          // For incoming streams, in case identify is on
+          this.muxer.on('stream', (conn) => {
```
can this `stream` event occurs multiple times?

jacobheun replied:
Yes. Once we have a muxed connection established it can open up any number of streams with that peer.

pgte commented on src/dialer.js:
```diff
@@ -0,0 +1,107 @@
+      connection.on('connected', () => connection.protect())
+      connection.on('private', () => connection.encrypt())
+      connection.on('encrypted', () => connection.upgrade())
+      connection.on('muxed', () => {
```
Can all these events occur more than once?

jacobheun replied:
No, they should only occur once, I'll update these.

pgte replied:
👍

vasco-santos commented on src/dialer.js:
```diff
@@ -0,0 +1,107 @@
+    const peerInfo = getPeerInfo(peer, _switch._peerBook)
+    const b58Id = peerInfo.id.toB58String()
+
+    log(`${_switch._peerInfo.id.toB58String().slice(0, 8)} dial request to ${b58Id.slice(0, 8)} with protocol ${protocol}`)
```
Verify if `protocol` exists and if not add `unknown` instead of allowing undefined

vasco-santos commented on src/index.js:
```diff
@@ -70,6 +79,43 @@ class Switch extends EE {
+      log('The switch has stopped')
+      this.emit('stopped')
+    })
+    this.state.on('error', (err) => this.emit('error', err))
```
Aiming to ease debuggability, I think we should log the error here as well.

vasco-santos commented on src/index.js:
```diff
@@ -167,7 +167,7 @@ class Switch extends EE {
   /**
    * If a muxed Connection exists for the given peer, it will be closed
-   * and its reference on the Switch will be removed.
+   * and its reference on the Switch and will be removed.
```
typo? I think it was correct before 🤔

vasco-santos commented on src/index.js:
```diff
@@ -197,6 +197,89 @@ class Switch extends EE {
+      this.transport.listen(ts, {}, null, cb)
+    }, (err) => {
+      if (err) {
+        return this.emit('error', err)
```
log the error?

vasco-santos commented on src/index.js:
```diff
@@ -197,6 +197,89 @@ class Switch extends EE {
+
+        conn.muxer.end((err) => {
+          // If OK things are fine, and someone just shut down
+          if (err && err.message !== 'Fatal error: OK') {
```
log the error?

vasco-santos commented on src/limit-dialer/queue.js:
```diff
@@ -37,15 +37,15 @@ class DialQueue {
     this._dialWithTimeout(transport, addr, (err, conn) => {
       if (err) {
-        log('work:error')
+        log(`${transport.constructor.name}:work:error`, err)
```
use `log.error` here

jacobheun commented:
After reviewing the crypto logic in switch as it currently is, we're not actually supporting multiple crypto modules on dial, only on listen. This needs to be corrected, but after doing some small changes in this branch I think it should wait until this is done as it's a significant change. That will enable us to review those changes more easily to ensure they're done properly.

Also, since secio is currently our only crypto library in libp2p this isn't a critical need.

daviddias commented on README.md:
```diff
@@ -26,16 +26,17 @@ libp2p-switch is used by [libp2p](https://github.com/libp2p/js-libp2p) but it ca
-  - [`switch.stop(callback)`](#swarmclosecallback)
-  - [`switch.connection`](#connection)
+  - [`switch.dial(peer, protocol, callback)`](#switchdialpeer-protocol-callback)
+  - [`switch.dialFSM(peer, protocol, callback)`](#switchdialfsmpeer-protocol-callback)
```
better just mimic js-libp2p and do dial and dialProtocol

jacobheun replied:
@diasdavid are you suggesting to add `dialProtocol` in addition to the new `dialFSM` or as a naming replacement? I think having `dial` handle calls with and without a protocol is reasonable overloading. Replacing it wouldn't be good as it has a different callback footprint to give users access to the ConnectionFSM for finer control/handling.  

I've exposed the same `dialFSM` method in the [libp2p branch](https://github.com/libp2p/js-libp2p/tree/feat/state-machine#libp2pdialfsmpeer-protocol-callback)

daviddias replied:
We don't need to block this PR for this. What I mean is that given that libp2p (the module) is just an extension to the libp2p-switch (the dialing machine) that adds it other features, it would be cool if they both expose the same dialing/hangup APIs.

jacobheun replied:
That makes sense. Ideally I'd like to deprecate `dialProtocol` in libp2p and just use `dial`. Having both feel unnecessary. I'll propose the change in an issue there and if `dialProtocol` ends up staying we can add that in here.

jacobheun commented:
@vasco-santos this should be ready for final review. I have this branch integrated in https://github.com/libp2p/js-libp2p/pull/257, which should also be ready for review.

vasco-santos commented on README.md:
```diff
@@ -26,16 +26,17 @@ libp2p-switch is used by [libp2p](https://github.com/libp2p/js-libp2p) but it ca
+  - [`switch.unhandle(protocol)`](#switchunhandleprotocol)
+  - [`switch.start(callback)`](#switchstartcallback)
+  - [`switch.stop(callback)`](#switchstopcallback)
+  - [`switch.connection`](#switchconnection)
```
Can we please alphabetize the API? I believe this helps when looking for documentation.

vasco-santos replied:
Can we please alphabetize the API? I believe this helps when looking for documentation.

vasco-santos commented on src/index.js:
```diff
@@ -197,6 +196,85 @@ class Switch extends EE {
+   * Issues a start on the Switch state.
+   *
+   * @param {function} callback deprecated: Listening for the `error` and `start` events are recommended
+   * @fires Switch#started
```
`@fires Switch#started` ?

vasco-santos commented on src/index.js:
```diff
@@ -197,6 +196,85 @@ class Switch extends EE {
+   * A listener that will turn off all running services and listeners
+   *
+   * @private
+   * @fires Switch#error
```
same as above

jacobheun commented:
@vasco-santos I updated the order in the readme and fixed the improper usage of jsdocs `@fires` on individual methods and updated those events on the Switch class itself.
